Inouts & sinks + contexts compile error

Using inout with sinks and contexts results in this error message:

```
$spicyz test.spicy -o test.hlto
In file included from /private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:6:
In file included from /Users/johanna/zeek/install/include/hilti/rt/libhilti.h:12:
In file included from /Users/johanna/zeek/install/include/hilti/rt/context.h:12:
In file included from /Users/johanna/zeek/install/include/hilti/rt/fiber.h:19:
/Users/johanna/zeek/install/include/hilti/rt/types/reference.h:720:31: error: no matching conversion for functional-style cast from '__hlt::Test::Share' to '__hlt::Test::Alert'
    return StrongReference<T>(T(std::forward<Args>(args)...));
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:573:60: note: in instantiation of function template specialization 'hilti::rt::reference::make_strong<__hlt::Test::Alert, __hlt::Test::Share &>' requested here
    (*(*__self).alertsink).connect(::hilti::rt::reference::make_strong<__hlt::Test::Alert>((*(*__self).__context)));
                                                           ^
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:99:9: note: candidate constructor not viable: no known conversion from '__hlt::Test::Share' to 'const Alert' for 1st argument
        Alert(const Alert&) = default;
        ^
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:100:9: note: candidate constructor not viable: no known conversion from '__hlt::Test::Share' to 'Alert' for 1st argument
        Alert(Alert&&) = default;
        ^
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:176:19: note: candidate constructor not viable: no known conversion from '__hlt::Test::Share' to '::hilti::rt::ValueReference<__hlt::Test::Share> &' for 1st argument
    inline Alert::Alert(::hilti::rt::ValueReference<__hlt::Test::Share>& __p_sh) : __p_sh(std::move(__p_sh)) {
                  ^
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:179:19: note: candidate constructor not viable: no known conversion from '__hlt::Test::Share' to 'std::optional<std::optional< ::hilti::rt::Vector<__hlt::Test::Alert_message>>>' for 1st argument
    inline Alert::Alert(std::optional<std::optional<::hilti::rt::Vector<__hlt::Test::Alert_message>>> alerts) : Alert() {
                  ^
/private/var/folders/50/z8y2zkwd4lsdl3qr2l4shn8c0000gn/T/Test_5e7bb92ea7bd7a6c-17aab9b4f987eead.cc:173:19: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
    inline Alert::Alert() {
                  ^
1 error generated.
[error] <Spicy support for Zeek>: JIT compilation failed
```

Source:
```
module Test;

type Share = unit {
};

public type Message = unit {
  %context = Share;

	sink alertsink;

	on %init {
		self.alertsink.connect(new Alert(self.context()));
	}
};

# note - this will mostly be garbage because it is encrypted.
public type Alert = unit(inout sh: Share) {
  alerts: Alert_message(sh)[];
};

type Alert_message = unit(sh: Share) {
  level: uint8; # &convert=AlertLevel($$);
  description: uint8; # &convert=AlertDescription($$);
};
```

(Previously mentioned as comment in #1711)

--- a/hilti/toolchain/include/ast/expressions/resolved-operator.h
+++ b/hilti/toolchain/include/ast/expressions/resolved-operator.h
@@ -40,7 +40,7 @@ public:
 
     std::string printSignature() const { return operator_::detail::printSignature(kind(), operands(), meta()); }
 
-    node::Properties properties() const final {
+    node::Properties properties() const override {
         auto p = node::Properties{{"kind", to_string(_operator->kind())}};
         return Expression::properties() + p;
     }

--- a/hilti/toolchain/include/ast/operators/common.h
+++ b/hilti/toolchain/include/ast/operators/common.h
@@ -10,9 +10,9 @@
 #include <hilti/ast/expressions/resolved-operator.h>
 #include <hilti/ast/forward.h>
 
-#define HILTI_NODE_OPERATOR(ns, cls)                                                                                   \
+#define HILTI_NODE_OPERATOR_CUSTOM_BASE(ns, cls, base)                                                                 \
     namespace ns {                                                                                                     \
-    class cls : public hilti::expression::ResolvedOperator {                                                           \
+    class cls : public base {                                                                                          \
     public:                                                                                                            \
         static cls* create(hilti::ASTContext* ctx, const hilti::Operator* op, hilti::QualifiedType* result,            \
                            const hilti::Expressions& operands, hilti::Meta meta) {                                     \
@@ -23,6 +23,8 @@
                                                                                                                        \
     private:                                                                                                           \
         cls(ASTContext* ctx, const hilti::Operator* op, QualifiedType* result, const Expressions& operands, Meta meta) \
-            : ResolvedOperator(ctx, NodeTags, op, result, operands, std::move(meta)) {}                                \
+            : base(ctx, NodeTags, op, result, operands, std::move(meta)) {}                                            \
     };                                                                                                                 \
     } // namespace ns
+
+#define HILTI_NODE_OPERATOR(ns, cls) HILTI_NODE_OPERATOR_CUSTOM_BASE(ns, cls, hilti::expression::ResolvedOperator)

--- a/hilti/toolchain/include/ast/operators/reference.h
+++ b/hilti/toolchain/include/ast/operators/reference.h
@@ -9,13 +9,39 @@
 
 namespace hilti::operator_ {
 
-HILTI_NODE_OPERATOR(strong_reference, Deref)
+namespace reference {
+
+/** Joint base class for all the references' `Deref` AST nodes. */
+class DerefBase : public expression::ResolvedOperator {
+public:
+    using expression::ResolvedOperator::ResolvedOperator;
+
+    /**
+     * Returns true if the operator has been marked as automatically created by
+     * the coercer.
+     */
+    auto isAutomaticCoercion() const { return _is_coercion; }
+
+    /** Marks the operators as automatically created by the coercer. */
+    void setIsAutomaticCoercion(bool is_coercion) { _is_coercion = is_coercion; }
+
+    node::Properties properties() const final {
+        auto p = node::Properties{{"auto", _is_coercion}};
+        return expression::ResolvedOperator::properties() + p;
+    }
+
+private:
+    bool _is_coercion = false;
+};
+} // namespace reference
+
+HILTI_NODE_OPERATOR_CUSTOM_BASE(strong_reference, Deref, reference::DerefBase)
 HILTI_NODE_OPERATOR(strong_reference, Equal)
 HILTI_NODE_OPERATOR(strong_reference, Unequal)
-HILTI_NODE_OPERATOR(weak_reference, Deref)
+HILTI_NODE_OPERATOR_CUSTOM_BASE(weak_reference, Deref, reference::DerefBase)
 HILTI_NODE_OPERATOR(weak_reference, Equal)
 HILTI_NODE_OPERATOR(weak_reference, Unequal)
-HILTI_NODE_OPERATOR(value_reference, Deref)
+HILTI_NODE_OPERATOR_CUSTOM_BASE(value_reference, Deref, reference::DerefBase)
 HILTI_NODE_OPERATOR(value_reference, Equal)
 HILTI_NODE_OPERATOR(value_reference, Unequal)
 

--- a/hilti/toolchain/src/compiler/codegen/operators.cc
+++ b/hilti/toolchain/src/compiler/codegen/operators.cc
@@ -525,7 +525,15 @@ struct Visitor : hilti::visitor::PreOrder {
         auto t = n->op0()->type()->type();
 
         if ( auto tv = t->tryAs<type::Type_>() ) {
-            auto args = util::join(tupleArguments(n, n->op1()), ", ");
+            auto ctor = n->op1()->as<expression::Ctor>()->ctor();
+
+            if ( auto x = ctor->tryAs<ctor::Coerced>() )
+                ctor = x->coercedCtor();
+
+            auto args = util::join(cg->compileCallArguments(ctor->as<ctor::Tuple>()->value(),
+                                                            tv->typeValue()->type()->parameters()),
+                                   ", ");
+
             result = fmt("::hilti::rt::reference::make_strong<%s>(%s)",
                          cg->compile(tv->typeValue(), codegen::TypeUsage::Ctor), args);
         }

--- a/hilti/toolchain/src/compiler/coercer.cc
+++ b/hilti/toolchain/src/compiler/coercer.cc
@@ -914,14 +914,22 @@ Expression* skipReferenceValue(Builder* builder, Expression* op) {
     if ( ! op->type()->type()->isReferenceType() )
         return op;
 
+    operator_::reference::DerefBase* deref = nullptr;
+
     if ( op->type()->type()->isA<type::ValueReference>() )
-        return *value_reference_deref->instantiate(builder, {op}, op->meta());
+        deref = static_cast<operator_::reference::DerefBase*>(
+            *value_reference_deref->instantiate(builder, {op}, op->meta()));
     else if ( op->type()->type()->isA<type::StrongReference>() )
-        return *strong_reference_deref->instantiate(builder, {op}, op->meta());
+        deref = static_cast<operator_::reference::DerefBase*>(
+            *strong_reference_deref->instantiate(builder, {op}, op->meta()));
     else if ( op->type()->type()->isA<type::WeakReference>() )
-        return *weak_reference_deref->instantiate(builder, {op}, op->meta());
+        deref = static_cast<operator_::reference::DerefBase*>(
+            *weak_reference_deref->instantiate(builder, {op}, op->meta()));
     else
         logger().internalError("unknown reference type");
+
+    deref->setIsAutomaticCoercion(true);
+    return deref;
 }
 
 static CoercedExpression _coerceExpression(Builder* builder, Expression* e, QualifiedType* src_, QualifiedType* dst_,

--- a/spicy/toolchain/src/compiler/codegen/codegen.cc
+++ b/spicy/toolchain/src/compiler/codegen/codegen.cc
@@ -92,6 +92,17 @@ struct VisitorPass1 : public visitor::MutatingPostOrder {
         auto new_n = builder()->ctorStruct(n->fields(), n->meta());
         replaceNode(n, new_n);
     }
+
+    void operator()(hilti::operator_::strong_reference::Deref* n) final {
+        if ( n->isAutomaticCoercion() ) {
+            // Revert any automatic derefs of units (or structs created from
+            // units) inserted by automatic coercion. We'll re-resolve them
+            // during HILTI compilation where needed for their value_refs.
+            auto sref = n->op0()->type()->type()->as<hilti::type::StrongReference>();
+            if ( auto dtype = sref->dereferencedType()->type(); dtype->isA<type::Unit>() || dtype->isOnHeap() )
+                replaceNode(n, n->op0(), "reverting strong_ref deref coercion");
+        }
+    }
 };
 
 // Visitor that runs repeatedly over the AST of a module until no further

--- a/tests/spicy/types/unit/context-inout.spicy
+++ b/tests/spicy/types/unit/context-inout.spicy
@@ -11,6 +11,11 @@ type X = unit(inout ctx: Ctx) {
 
 public type Y = unit {
 	%context = Ctx;
+
+	on %init {
+		new X(self.context()); # make sure this compiles as well (regression test for #1719)
+	}
+
 	: X(self.context());
 
 	on %done() { print self.context(); }

